### PR TITLE
Implement printing of stack traces on LLVM segfaults and aborts

### PR DESCRIPTION
**Pull request by in42**
Implement #79153

Based on discussion, try to extend the rust_backtrace=1 feature to handle segfault or aborts in the llvm backend

### CODE REVIEW

**Comment by tmandry** [comment]
r? @tmandry 

Just saw this go by. I'll give it a full review next week. Thanks!

**Comment by tmandry** [comment]
@in42 It looks like you'll want a `#[cfg(unix)]` around this specific implementation, at least.

**Comment by in42** [comment]
This is a very basic implementation, the stack trace printed is barely readable, I will improve it this weekend.

**Comment by in42** [comment]
Due to personal reasons, I will not be able to work on this anymore.

**Comment by tmandry** [comment]
> This is a very basic implementation, the stack trace printed is barely readable, I will improve it this weekend.

That's why I was thinking we might want to use LLVM's hooks when using it as a backend; we already get all the nice stuff "for free." Not sure if it's actually feasible and reachable through a public API, though I remember thinking it was.

**Comment by in42** [comment]
Yeah basically there is no point in re implementing all of the functionality in rust - will look into how the llvm functions can be used.

**Comment by Aaron1011** [comment]
You can also use `std::backtrace::Backtrace` instead of the libc backtrace function.

**Comment by in42** [comment]
> You can also use `std::backtrace::Backtrace` instead of the libc backtrace function.

Ok will look into these

**Comment by in42** [comment]
> You can also use `std::backtrace::Backtrace` instead of the libc backtrace function.

I tried to use it in the following way:
```
#[cfg(unix)]
fn print_stack_trace(_: libc::c_int) {
    let backtrace = std::backtrace::Backtrace::capture();
    println!("{}", backtrace);
}
```
and run the compiler like this: 
```
$ RUST_BACKTRACE=1 build/x86_64-unknown-linux-gnu/stage1/bin/rustc -Aunused -g a.rs
[1]    2202274 segmentation fault (core dumped)  RUST_BACKTRACE=1 build/x86_64-unknown-linux-gnu/stage1/bin/rustc -Aunused -g
```
but no backtrace is getting printed.

**Comment by in42** [comment]
Using `libc::backtrace` function, the backtrace being printed was of this form:
```
$ build/x86_64-unknown-linux-gnu/stage1/bin/rustc -Aunused -g a.rs
[                                                                                                                                                                                                       
    0x00007f31c3ad5cfa,
    0x00007f31c312a6a0,                                                                                                                                                                                 
    0x00007f31c0cbc312,                                                                                                                                                                                 
    0x00007f31c0cbb619,
    ....
]
```

**Comment by tmandry** [comment]
Hm, maybe if we get something like #77384 merged we could use that.

**Comment by in42** [comment]
> Hm, maybe if we get something like #77384 merged we could use that.

Yeah, that will be better, I won’t have to implement the backtrace myself. 😃

**Comment by tmandry** [comment]
Talking with @yaahc gave me another idea: call the panic hook! The [default panic hook](https://github.com/rust-lang/rust/blob/16b805713c4cdb967263b9d59634fddea4372b65/library/std/src/panicking.rs#L180) prints a backtrace without allocating. If we call that we'll get a backtrace, and then we can abort.

The compiler does install another panic hook, but saves a handle to the default hook [here](https://github.com/rust-lang/rust/blob/16b805713c4cdb967263b9d59634fddea4372b65/compiler/rustc_driver/src/lib.rs#L1178). So we can call it directly from librustc_driver.

Worth a shot, I think..

**Comment by in42** [comment]
> Talking with @yaahc gave me another idea: call the panic hook! The [default panic hook](https://github.com/rust-lang/rust/blob/16b805713c4cdb967263b9d59634fddea4372b65/library/std/src/panicking.rs#L180) prints a backtrace without allocating. If we call that we'll get a backtrace, and then we can abort.
> 
> The compiler does install another panic hook, but saves a handle to the default hook [here](https://github.com/rust-lang/rust/blob/16b805713c4cdb967263b9d59634fddea4372b65/compiler/rustc_driver/src/lib.rs#L1178). So we can call it directly from librustc_driver.
> 
> Worth a shot, I think..

ok will try this

**Comment by tmandry** [comment]
@in42 following up here, have you had a chance to try the panic hook approach?

**Comment by in42** [comment]
> @in42 following up here, have you had a chance to try the panic hook approach?

Sorry was a little busy these last few weeks, will try this weekend.

**Comment by in42** [comment]
I am having a little problem populating the `PanicInfo` struct which needs to be sent to the `DEFAULT_HOOK`. I am trying this:
`let info = panic::PanicInfo::internal_constructor(Some(std::fmt::Arguments { pieces: "hello", fmt: "hello", args: "hello" }), Location::caller());`. 
This is giving a compiler error.
Is there an easy to create this struct? Calling `PanicInfo::internal_constructor` doesn't seem right.
Also I am not sure how to create the `std::fmt::Arguments` struct.

**Comment by in42** [comment]
In the meantime, I can look into properly printing the backtrace with function names (currently only the addresses are getting printed) by copying what llvm is doing i.e. the approach I was following earlier.

**Comment by in42** [comment]
Hi, I had a doubt, will the `DEFAULT HOOK` able to print backtrace if the functions are not from rust but from cpp part of the code? Also I am having a little trouble understanding how to populate `PanicInfo` struct to send to the `DEFAULT_HOOK`.

**Comment by in42** [comment]
On the other hand, I found out I just need to port this function to rust: https://github.com/rust-lang/llvm-project/blob/40618a2340ac2939b3dc320a9fe1133fa5caa345/llvm/lib/Support/Signals.cpp#L108-L221, and I will able to print backtrace with function names based on the approach I was following earlier. So simultaneously, I will also try to proceed along this approach.

**Comment by in42** [comment]
One thing I can do for now is check if llvm-symbolizer is in PATH then use it to print symbolized backtraces.

**Comment by in42** [comment]
Also @bjorn3 suggested this [crate](https://crates.io/crates/backtrace) on zulip.

**Comment by in42** [comment]
@bjorn3 suggested [addr2line](https://crates.io/crates/addr2line) crate on zulip.

File: compiler/rustc_driver/src/lib.rs [hunk 1]
```diff
@@ -1323,9 +1323,46 @@ pub fn init_env_logger(env: &str) {
     tracing::subscriber::set_global_default(subscriber).unwrap();
 }
 
+#[cfg(unix)]
+fn print_stack_trace(_: libc::c_int) {
+    unsafe {
+        static mut STACK_TRACE: [*mut libc::c_void; 256] = [std::ptr::null_mut(); 256];
+        let _depth = libc::backtrace(STACK_TRACE.as_mut_ptr(), 256);
```
Comment (bjorn3) [inline compiler/rustc_driver/src/lib.rs:1330]: Libc's `backtrace` function does a naive walk of frame pointers. Rustc and LLVM are normally compiled with frame pointers omitted, meaning that only DWARF based unwinding is reliable.

https://www.gnu.org/software/libc/manual/html_node/Backtraces.html

> [...] frame pointer elimination will stop `backtrace` from interpreting the stack contents correctly.

In addition it is not safe to call it from asynchronous signals like SIGSEGV.

> [...] AS-Unsafe init heap dlopen plugin lock [...]

Comment (in42) [reply compiler/rustc_driver/src/lib.rs:1330]: Yeah I noticed inside the llvm code, using the backtrace c api was one of the ways they were generating the backtrace. I will try to add the other ways too.

**Comment by tmandry** [comment]
> Hi, I had a doubt, will the `DEFAULT HOOK` able to print backtrace if the functions are not from rust but from cpp part of the code? Also I am having a little trouble understanding how to populate the `PanicInfo` struct to send to the `DEFAULT_HOOK`.

Sorry, forgot about that. It's a bit hacky, but it seems like the most straightforward way to follow this approach would be, roughly:

- reinstall the default hook (change the type of `DEFAULT_HOOK` so we can get it back somehow.. probably wrap it in `Mutex<Option<..>>`)
- actually panic.

We can exit the process afterward so don't have to worry about cleaning up.

File: compiler/rustc_driver/src/lib.rs (hunk 1 above)
Comment (in42) [reply compiler/rustc_driver/src/lib.rs:1330]: @bjorn3 [Here](https://github.com/rust-lang/llvm-project/blob/62a1ddde22c267249eda72184520a21ad2052f0b/llvm/lib/Support/Unix/Signals.inc#L558-L628) you can see, the glibc `backtrace` function is being used in llvm to print the backtrace. The alternative method being used is unwindBacktrace, but I think it is for ARM systems.

**Comment by in42** [comment]
Hi, in the latest commit I have been able to print out symbolized backtraces. I am using the `glibc` functions `backtrace` and `backtrace_symbols_fd`.

Sample backtrace:
```
$ build/x86_64-unknown-linux-gnu/stage1/bin/rustc -Aunused -g a.rs
/home/vikram/workspaces/rustc/rust/build/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_driver-1f164c31563ba402.so(+0x679bc3)[0x7f6afcab1bc3]
/usr/lib/libc.so.6(+0x3d6a0)[0x7f6afc1066a0]
/home/vikram/workspaces/rustc/rust/build/x86_64-unknown-linux-gnu/stage1/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so(_ZN4llvm16DwarfCompileUnit24constructVariableDIEImplERKNS_11DbgVariableEb+0xf2)[0x7f6af9c98312]
....

**Comment by in42** [comment]
Ok as far as I have understood the panic hook approach, the other panic hook the compiler installs calls the default hook [here](https://github.com/rust-lang/rust/blob/16b805713c4cdb967263b9d59634fddea4372b65/compiler/rustc_driver/src/lib.rs#L1193), so I do not reinstall it. I can actually just panic, and the default hook will be called. So I tried that using this code:
```
fn print_stack_trace(_: libc::c_int) {
    panic!("help");
    // unsafe {
    //     static mut STACK_TRACE: [*mut libc::c_void; 256] = [std::ptr::null_mut(); 256];
    //     let depth = libc::backtrace(STACK_TRACE.as_mut_ptr(), 256);
    //     if depth == 0 {
    //         return;
    //     }
    //     backtrace_symbols_fd(STACK_TRACE.as_ptr(), depth, 2);
    // }
}
```
and ran the compiler like this:
```
$ RUST_BACKTRACE=1 build/x86_64-unknown-linux-gnu/stage1/bin/rustc -Aunused -g a.rs
thread '<unnamed>' panicked at 'help', compiler/rustc_driver/src/lib.rs:1333:5
stack backtrace:
[1]    3509283 segmentation fault (core dumped)  RUST_BACKTRACE=1 build/x86_64-unknown-linux-gnu/stage1/bin/rustc -Aunused -g
```
However, as you can see, the backtrace is not getting printed.

**Comment by in42** [comment]
In the latest commit, I have been able to print the symbolized backtrace. I am using the glibc function `backtrace_symbols_fd` for it.
Sample backtrace:
```
/home/vikram/workspaces/rustc/rust/build/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_driver-1f164c31563ba402.so(+0x679bc3)[0x7f6afcab1bc3]
/usr/lib/libc.so.6(+0x3d6a0)[0x7f6afc1066a0]
/home/vikram/workspaces/rustc/rust/build/x86_64-unknown-linux-gnu/stage1/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so(_ZN4llvm16DwarfCompileUnit24constructVariableDIEImplERKNS_11DbgVa
riableEb+0xf2)[0x7f6af9c98312]
/home/vikram/workspaces/rustc/rust/build/x86_64-unknown-linux-gnu/stage1/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so(_ZN4llvm16DwarfCompileUnit22createScopeChildrenDIEEPNS_12LexicalS
copeERNS_15SmallVectorImplIPNS_3DIEEEEPb+0xe59)[0x7f6af9c97619]
```

**Comment by in42** [comment]
I will also try to symbolize the backtrace using `addr2line`.

File: compiler/rustc_driver/src/lib.rs (hunk 1 above)
Comment (in42) [reply compiler/rustc_driver/src/lib.rs]: @bjorn3 [Here](https://github.com/rust-lang/llvm-project/blob/62a1ddde22c267249eda72184520a21ad2052f0b/llvm/lib/Support/Unix/Signals.inc#L565), llvm uses the `backtrace` function to get the backtrace. If it fails it tries to use `_Unwind_Backtrace`.

File: compiler/rustc_driver/src/lib.rs [hunk 2]
```diff
@@ -1323,9 +1323,54 @@ pub fn init_env_logger(env: &str) {
     tracing::subscriber::set_global_default(subscriber).unwrap();
 }
 
+extern "C" {
+    // Only available in glibc
```
Comment (tmandry) [inline compiler/rustc_driver/src/lib.rs:1327]: nit: Not sure if this is true, it seems available on macOS which I don't think uses glibc?

**Review by tmandry — APPROVED** [review]
Nice! This is going to be really helpful for debugging crashes "in the field".

@bjorn3 did you want to take a look before merging?

**Comment by in42** [comment]
> Nice! This is going to be really helpful for debugging crashes "in the field".
> 
> @bjorn3 did you want to take a look before merging?

There is one problem that this won't work on windows for example. Also, the mingw build is failing due to presence of dead code which I have to fix.

File: compiler/rustc_driver/src/lib.rs [hunk 3]
```diff
@@ -1323,9 +1323,54 @@ pub fn init_env_logger(env: &str) {
...
+                as *mut libc::c_void;
+        alt_stack.ss_size = ALT_STACK_SIZE;
+        libc::sigaltstack(&mut alt_stack, std::ptr::null_mut());
+
+        let mut sa: libc::sigaction = std::mem::zeroed();
+        sa.sa_sigaction =
+            print_stack_trace as fn(libc::c_int) as *mut libc::c_void as libc::sighandler_t;
+        sa.sa_flags = libc::SA_NODEFER | libc::SA_RESETHAND | libc::SA_ONSTACK;
+        libc::sigemptyset(&mut sa.sa_mask);
+        libc::sigaction(libc::SIGSEGV, &sa, std::ptr::null_mut());
```
Comment (tmandry) [inline compiler/rustc_driver/src/lib.rs:1361]: I think LLVM also issues `SIGABRT` when assertions are enabled, and I think a backtrace would be useful there too.

That being said, I don't think we should print this when the compiler dies due to a Rust panic, so if that happens then please don't add that signal :)

**Review by tmandry — COMMENTED** [review]
Okay, left a few more comments after a closer look :)

**Comment by tmandry** [comment]
> There is one problem that this won't work on windows for example. Also, the mingw build is failing due to presence of dead code which I have to fix.

I'm okay with it not working on windows for now. This is definitely an improvement and we can always expand it in follow-up PRs if we choose to.

FYI, I do expect some CI configurations might break (I'm glad mingw is in the smoke tests since it's a great edge case detector). So it might take a few tries to actually get this merged.

**Comment by crlf0710** [comment]
@in42 Ping from triage, is this ready for review? If so, would you mind removing `[WIP]` from PR title?

**Comment by in42** [comment]
> @in42 Ping from triage, is this ready for review? If so, would you mind removing `[WIP]` from PR title?

Yes, it is.

**Comment by Dylan-DPC** [comment]
@tmandry this is ready for review

**Comment by tmandry** [comment]
I took the liberty of squashing the intermediate commits and adding a few small fixes. With that I think this PR is ready to land.

@bors r+

**Pull request merged**